File: python/xenarch/router.py
"""Facilitator Router — selects an x402 facilitator per payment.

Implements `Information/design/facilitator-router.md`. The Router picks a
ranked fallback list of facilitators for each payment based on a weighted
score (fee, gas sponsorship, spec compliance, observed uptime, observed
latency, publisher preference). Failed settlements are recorded as
passive health signals; a facilitator that exceeds the failure threshold
within the rolling window is circuit-broken for the cooldown period.

The Router only selects from facilitators it was constructed with (its
"registered stack"). A publisher's pay.json may rank a SUBSET of the
registered stack; URLs not in the stack are ignored. This keeps the
Router from making outbound calls to arbitrary URLs sourced from
publisher metadata.
"""

from __future__ import annotations

import time
from collections import deque
from dataclasses import dataclass, field
from decimal import Decimal
from enum import Enum
from typing import Callable, Sequence


# §9 — default facilitator stack. Coinbase is configurable, never default.
DEFAULT_FACILITATOR_STACK: tuple["FacilitatorConfig", ...]
# (Defined after FacilitatorConfig.)

# §6.1 circuit breaker defaults.
DEFAULT_FAILURE_THRESHOLD = 5
DEFAULT_FAILURE_WINDOW_S = 60.0
DEFAULT_COOLDOWN_S = 300.0

# §5.2 weights — sum to 1.0.
DEFAULT_WEIGHTS: dict[str, float] = {
    "fee": 0.30,
    "gas": 0.25,
    "spec": 0.10,
    "uptime": 0.15,
    "latency": 0.10,
    "preference": 0.10,
}


class HealthState(str, Enum):
    HEALTHY = "healthy"
    DEGRADED = "degraded"
    UNHEALTHY = "unhealthy"


@dataclass(frozen=True)
class FacilitatorConfig:
    """Static facilitator metadata. Mutable runtime state lives in `_FacilitatorState`."""

    name: str
    url: str
    spec_version: str = "v2"
    fee_bps: int = 0
    gas_sponsored: bool = True
    supported_chains: frozenset[str] = frozenset({"base"})
    supported_assets: frozenset[str] = frozenset({"USDC"})


DEFAULT_FACILITATOR_STACK = (
    FacilitatorConfig(
        name="PayAI",
        url="https://facilitator.payai.network",
    ),
    FacilitatorConfig(
        name="xpay",
        url="https://facilitator.xpay.dev",
    ),
    FacilitatorConfig(
        name="Ultravioleta DAO",
        url="https://x402.ultravioletadao.xyz",
    ),
    FacilitatorConfig(
        name="x402.rs",
        url="https://x402.rs",
    ),
)


@dataclass
class PaymentContext:
    chain: str = "base"
    asset: str = "USDC"
    amount_usd: Decimal | None = None
    publisher_max_fee_bps: int | None = None


@dataclass
class _FacilitatorState:
    config: FacilitatorConfig
    failures: deque[float] = field(default_factory=lambda: deque(maxlen=200))
    successes: deque[float] = field(default_factory=lambda: deque(maxlen=200))
    latency_samples_ms: deque[float] = field(default_factory=lambda: deque(maxlen=50))
    cooldown_until: float = 0.0

    def health(self, now: float, threshold: int, window_s: float) -> HealthState:
        if now < self.cooldown_until:
            return HealthState.UNHEALTHY
        recent = sum(1 for t in self.failures if now - t < window_s)
        if recent >= threshold:
            return HealthState.UNHEALTHY
        if recent > 0:
            return HealthState.DEGRADED
        return HealthState.HEALTHY


class Router:
    """Selects facilitators per payment, tracks passive health, falls back on failure."""

    def __init__(
        self,
        facilitators: Sequence[FacilitatorConfig] | None = None,
        weights: dict[str, float] | None = None,
        failure_threshold: int = DEFAULT_FAILURE_THRESHOLD,
        failure_window_s: float = DEFAULT_FAILURE_WINDOW_S,
        cooldown_s: float = DEFAULT_COOLDOWN_S,
        clock: Callable[[], float] = time.monotonic,
    ) -> None:
        configs = (
            list(facilitators) if facilitators is not None else list(DEFAULT_FACILITATOR_STACK)
        )
        if not configs:
            raise ValueError("Router requires at least one facilitator")
        # Preserve insertion order; dict is ordered in Py3.7+.
        self._states: dict[str, _FacilitatorState] = {
            c.url: _FacilitatorState(config=c) for c in configs
        }
        self._weights = dict(weights) if weights is not None else dict(DEFAULT_WEIGHTS)
        self._failure_threshold = failure_threshold
        self._failure_window_s = failure_window_s
        self._cooldown_s = cooldown_s
        self._clock = clock
        self._last_used_url: str | None = None

    @property
    def registered(self) -> list[FacilitatorConfig]:
        return [s.config for s in self._states.values()]

    def select(
        self,
        ctx: PaymentContext | None = None,
        publisher_facilitators: Sequence[str] | None = None,
        max_results: int = 3,
    ) -> list[FacilitatorConfig]:
        """Return a ranked fallback list. Empty list if nothing eligible."""
        ctx = ctx or PaymentContext()
        now = self._clock()

        candidates: list[_FacilitatorState] = []
        for state in self._states.values():
            cfg = state.config
            if ctx.chain not in cfg.supported_chains:
                continue
            if ctx.asset not in cfg.supported_assets:
                continue
            if (
                ctx.publisher_max_fee_bps is not None
                and cfg.fee_bps > ctx.publisher_max_fee_bps
            ):
                continue
            if state.health(now, self._failure_threshold, self._failure_window_s) == HealthState.UNHEALTHY:
                continue
            candidates.append(state)

        if not candidates:
            return []

        publisher_rank: dict[str, int] = {}
        if publisher_facilitators:
            i = 0
            for url in publisher_facilitators:
                if url in self._states and url not in publisher_rank:
                    publisher_rank[url] = i
                    i += 1

        scored = [(self._score(s, now, publisher_rank), s) for s in candidates]
        scored.sort(key=lambda pair: pair[0], reverse=True)

        # §5.4 tie-break: within 1%, prefer different from last used.
        if len(scored) >= 2 and self._last_used_url:
            top_score, top_state = scored[0]
            next_score, next_state = scored[1]
            denom = max(abs(top_score), 1e-9)
            if abs(top_score - next_score) / denom < 0.01:
                if top_state.config.url == self._last_used_url:
                    scored[0], scored[1] = scored[1], scored[0]

        return [s.config for _, s in scored[:max_results]]

    def record_success(self, url: str, latency_ms: float | None = None) -> None:
        state = self._states.get(url)
        if state is None:
            return
        now = self._clock()
        state.successes.append(now)
        if latency_ms is not None:
            state.latency_samples_ms.append(latency_ms)
        self._last_used_url = url

    def record_failure(self, url: str) -> None:
        state = self._states.get(url)
        if state is None:
            return
        now = self._clock()
        state.failures.append(now)
        recent = sum(1 for t in state.failures if now - t < self._failure_window_s)
        if recent >= self._failure_threshold:
            state.cooldown_until = now + self._cooldown_s

    def health(self, url: str) -> HealthState | None:
        state = self._states.get(url)
        if state is None:
            return None
        return state.health(
            self._clock(), self._failure_threshold, self._failure_window_s
        )

    def _score(
        self,
        state: _FacilitatorState,
        now: float,
        publisher_rank: dict[str, int],
    ) -> float:
        cfg = state.config
        w = self._weights
        # Fee: 0bps → 1.0, 100bps+ → 0.0. Linear in between.
        fee_component = 1.0 - min(cfg.fee_bps / 100.0, 1.0)
        gas_component = 1.0 if cfg.gas_sponsored else 0.0
        spec_component = 1.0 if cfg.spec_version == "v2" else 0.5
        uptime_component = self._uptime_component(state, now)
        latency_component = self._latency_component(state)
        if cfg.url in publisher_rank and publisher_rank:
            n = max(len(publisher_rank), 1)
            # Rank 0 (top) → 1.0; last rank → ~0.0.
            preference_component = 1.0 - (publisher_rank[cfg.url] / n)
        else:
            preference_component = 0.0
        return (
            w["fee"] * fee_component
            + w["gas"] * gas_component
            + w["spec"] * spec_component
            + w["uptime"] * uptime_component
            + w["latency"] * latency_component
            + w["preference"] * preference_component
        )

    @staticmethod
    def _uptime_component(state: _FacilitatorState, now: float) -> float:
        # Spec calls for 7-day rolling; deques are bounded at 200 events,
        # which is approximate but cheap. Untested = optimistic.
        window_s = 7 * 24 * 3600.0
        s = sum(1 for t in state.successes if now - t < window_s)
        f = sum(1 for t in state.failures if now - t < window_s)
        total = s + f
        if total == 0:
            return 1.0
        return s / total

    @staticmethod
    def _latency_component(state: _FacilitatorState) -> float:
        if not state.latency_samples_ms:
            return 1.0
        # p95 inverted, clamped to [500ms, 5000ms].
        sorted_samples = sorted(state.latency_samples_ms)
        p95_idx = max(0, int(len(sorted_samples) * 0.95) - 1)
        p95 = sorted_samples[p95_idx]
        if p95 <= 500:
            return 1.0
        if p95 >= 5000:
            return 0.0
        return 1.0 - (p95 - 500) / 4500

File: cli/src/lib/router.ts
/**
 * Facilitator Router — selects an x402 facilitator per payment.
 *
 * TypeScript port of `xenarch-sdks/python/xenarch/router.py`. Implements
 * `Information/design/facilitator-router.md`. The Router picks a ranked
 * fallback list of facilitators for each payment based on a weighted score
 * (fee, gas sponsorship, spec compliance, observed uptime, observed latency,
 * publisher preference). Failed settlements are recorded as passive health
 * signals; a facilitator that exceeds the failure threshold within the
 * rolling window is circuit-broken for the cooldown period.
 *
 * The Router only selects from facilitators it was constructed with (its
 * "registered stack"). A publisher's pay.json may rank a SUBSET of the
 * registered stack; URLs not in the stack are ignored. This keeps the
 * Router from making outbound calls to arbitrary URLs sourced from
 * publisher metadata.
 */

export interface FacilitatorConfig {
  name: string;
  url: string;
  /** spec_version: "v2" preferred, "v1" supported but downscored. */
  specVersion?: string;
  /** Fee in basis points; 0 means free. */
  feeBps?: number;
  /** Whether the facilitator sponsors gas for the on-chain tx. */
  gasSponsored?: boolean;
  /** Set of chain ids the facilitator supports (e.g. "base"). */
  supportedChains?: Set<string>;
  /** Set of asset symbols (e.g. "USDC"). */
  supportedAssets?: Set<string>;
}

export interface PaymentContext {
  /** Default "base". */
  chain?: string;
  /** Default "USDC". */
  asset?: string;
  /** Decimal as string. Informational only — does not filter. */
  amountUsd?: string;
  /** Publisher-imposed max fee in bps. */
  publisherMaxFeeBps?: number;
}

export type HealthState = "healthy" | "degraded" | "unhealthy";

export interface RouterOptions {
  weights?: Partial<Record<ScoreAxis, number>>;
  failureThreshold?: number;
  failureWindowS?: number;
  cooldownS?: number;
  /** Override clock for tests; default `() => Date.now() / 1000`. */
  clock?: () => number;
}

type ScoreAxis = "fee" | "gas" | "spec" | "uptime" | "latency" | "preference";

// §6.1 circuit breaker defaults.
export const DEFAULT_FAILURE_THRESHOLD = 5;
export const DEFAULT_FAILURE_WINDOW_S = 60;
export const DEFAULT_COOLDOWN_S = 300;

// §5.2 weights — sum to 1.0.
export const DEFAULT_WEIGHTS: Record<ScoreAxis, number> = {
  fee: 0.3,
  gas: 0.25,
  spec: 0.1,
  uptime: 0.15,
  latency: 0.1,
  preference: 0.1,
};

// §9 — default facilitator stack. Coinbase is configurable, never default.
export const DEFAULT_FACILITATOR_STACK: FacilitatorConfig[] = [
  { name: "PayAI", url: "https://facilitator.payai.network" },
  { name: "xpay", url: "https://facilitator.xpay.dev" },
  { name: "Ultravioleta DAO", url: "https://x402.ultravioletadao.xyz" },
  { name: "x402.rs", url: "https://x402.rs" },
];

/** Bounded ring buffer for timestamps + latency samples. */
class RingBuffer<T> {
  private items: T[] = [];
  constructor(private readonly cap: number) {}

  push(v: T): void {
    this.items.push(v);
    if (this.items.length > this.cap) {
      this.items.shift();
    }
  }

  toArray(): T[] {
    return this.items.slice();
  }

  get length(): number {
    return this.items.length;
  }
}

interface ResolvedConfig {
  name: string;
  url: string;
  specVersion: string;
  feeBps: number;
  gasSponsored: boolean;
  supportedChains: Set<string>;
  supportedAssets: Set<string>;
}

function resolve(c: FacilitatorConfig): ResolvedConfig {
  return {
    name: c.name,
    url: c.url,
    specVersion: c.specVersion ?? "v2",
    feeBps: c.feeBps ?? 0,
    gasSponsored: c.gasSponsored ?? true,
    supportedChains: c.supportedChains ?? new Set(["base"]),
    supportedAssets: c.supportedAssets ?? new Set(["USDC"]),
  };
}

interface FacilitatorState {
  config: ResolvedConfig;
  failures: RingBuffer<number>;
  successes: RingBuffer<number>;
  latencyMs: RingBuffer<number>;
  cooldownUntil: number;
}

function defaultClock(): number {
  return Date.now() / 1000;
}

export class Router {
  private readonly states: Map<string, FacilitatorState>;
  private readonly weights: Record<ScoreAxis, number>;
  private readonly failureThreshold: number;
  private readonly failureWindowS: number;
  private readonly cooldownS: number;
  private readonly clock: () => number;
  private lastUsedUrl: string | null = null;

  constructor(facilitators?: FacilitatorConfig[], opts: RouterOptions = {}) {
    const configs =
      facilitators !== undefined
        ? facilitators.slice()
        : DEFAULT_FACILITATOR_STACK.slice();
    if (configs.length === 0) {
      throw new Error("Router requires at least one facilitator");
    }
    // Map preserves insertion order in JS.
    this.states = new Map();
    for (const c of configs) {
      const resolved = resolve(c);
      this.states.set(resolved.url, {
        config: resolved,
        failures: new RingBuffer<number>(200),
        successes: new RingBuffer<number>(200),
        latencyMs: new RingBuffer<number>(50),
        cooldownUntil: 0,
      });
    }
    this.weights = { ...DEFAULT_WEIGHTS, ...(opts.weights ?? {}) };
    this.failureThreshold = opts.failureThreshold ?? DEFAULT_FAILURE_THRESHOLD;
    this.failureWindowS = opts.failureWindowS ?? DEFAULT_FAILURE_WINDOW_S;
    this.cooldownS = opts.cooldownS ?? DEFAULT_COOLDOWN_S;
    this.clock = opts.clock ?? defaultClock;
  }

  /** Snapshot of registered configs (mutating the result is harmless). */
  get registered(): FacilitatorConfig[] {
    return Array.from(this.states.values()).map((s) => ({
      name: s.config.name,
      url: s.config.url,
      specVersion: s.config.specVersion,
      feeBps: s.config.feeBps,
      gasSponsored: s.config.gasSponsored,
      supportedChains: new Set(s.config.supportedChains),
      supportedAssets: new Set(s.config.supportedAssets),
    }));
  }

  /** Return a ranked fallback list. Empty list if nothing eligible. */
  select(
    ctx: PaymentContext = {},
    publisherFacilitators?: string[],
    maxResults: number = 3,
  ): FacilitatorConfig[] {
    const chain = ctx.chain ?? "base";
    const asset = ctx.asset ?? "USDC";
    const now = this.clock();

    const candidates: FacilitatorState[] = [];
    for (const state of this.states.values()) {
      const cfg = state.config;
      if (!cfg.supportedChains.has(chain)) continue;
      if (!cfg.supportedAssets.has(asset)) continue;
      if (
        ctx.publisherMaxFeeBps !== undefined &&
        cfg.feeBps > ctx.publisherMaxFeeBps
      )
        continue;
      if (this.healthState(state, now) === "unhealthy") continue;
      candidates.push(state);
    }

    if (candidates.length === 0) return [];

    const publisherRank = new Map<string, number>();
    if (publisherFacilitators && publisherFacilitators.length > 0) {
      let i = 0;
      for (const url of publisherFacilitators) {
        if (this.states.has(url) && !publisherRank.has(url)) {
          publisherRank.set(url, i);
          i += 1;
        }
      }
    }

    const scored = candidates.map((s) => ({
      score: this.score(s, now, publisherRank),
      state: s,
    }));
    scored.sort((a, b) => b.score - a.score);

    // §5.4 tie-break: within 1%, prefer different from last used.
    if (scored.length >= 2 && this.lastUsedUrl) {
      const top = scored[0];
      const next = scored[1];
      const denom = Math.max(Math.abs(top.score), 1e-9);
      if (
        Math.abs(top.score - next.score) / denom < 0.01 &&
        top.state.config.url === this.lastUsedUrl
      ) {
        scored[0] = next;
        scored[1] = top;
      }
    }

    return scored.slice(0, maxResults).map((p) => ({
      name: p.state.config.name,
      url: p.state.config.url,
      specVersion: p.state.config.specVersion,
      feeBps: p.state.config.feeBps,
      gasSponsored: p.state.config.gasSponsored,
      supportedChains: new Set(p.state.config.supportedChains),
      supportedAssets: new Set(p.state.config.supportedAssets),
    }));
  }

  recordSuccess(url: string, latencyMs?: number): void {
    const state = this.states.get(url);
    if (state === undefined) return;
    const now = this.clock();
    state.successes.push(now);
    if (latencyMs !== undefined) state.latencyMs.push(latencyMs);
    this.lastUsedUrl = url;
  }

  recordFailure(url: string): void {
    const state = this.states.get(url);
    if (state === undefined) return;
    const now = this.clock();
    state.failures.push(now);
    let recent = 0;
    for (const t of state.failures.toArray()) {
      if (now - t < this.failureWindowS) recent += 1;
    }
    if (recent >= this.failureThreshold) {
      state.cooldownUntil = now + this.cooldownS;
    }
  }

  health(url: string): HealthState | null {
    const state = this.states.get(url);
    if (state === undefined) return null;
    return this.healthState(state, this.clock());
  }

  private healthState(state: FacilitatorState, now: number): HealthState {
    if (now < state.cooldownUntil) return "unhealthy";
    let recent = 0;
    for (const t of state.failures.toArray()) {
      if (now - t < this.failureWindowS) recent += 1;
    }
    if (recent >= this.failureThreshold) return "unhealthy";
    if (recent > 0) return "degraded";
    return "healthy";
  }

  private score(
    state: FacilitatorState,
    now: number,
    publisherRank: Map<string, number>,
  ): number {
    const cfg = state.config;
    const w = this.weights;
    const fee = 1.0 - Math.min(cfg.feeBps / 100.0, 1.0);
    const gas = cfg.gasSponsored ? 1.0 : 0.0;
    const spec = cfg.specVersion === "v2" ? 1.0 : 0.5;
    const uptime = this.uptimeComponent(state, now);
    const latency = this.latencyComponent(state);
    let preference = 0.0;
    if (publisherRank.size > 0 && publisherRank.has(cfg.url)) {
      const rank = publisherRank.get(cfg.url)!;
      const n = Math.max(publisherRank.size, 1);
      preference = 1.0 - rank / n;
    }
    return (
      w.fee * fee +
      w.gas * gas +
      w.spec * spec +
      w.uptime * uptime +
      w.latency * latency +
      w.preference * preference
    );
  }

  private uptimeComponent(state: FacilitatorState, now: number): number {
    // Spec calls for 7-day rolling; ring buffers are bounded at 200 events,
    // which is approximate but cheap. Untested = optimistic.
    const windowS = 7 * 24 * 3600;
    let s = 0;
    for (const t of state.successes.toArray()) {
      if (now - t < windowS) s += 1;
    }
    let f = 0;
    for (const t of state.failures.toArray()) {
      if (now - t < windowS) f += 1;
    }
    const total = s + f;
    if (total === 0) return 1.0;
    return s / total;
  }

  private latencyComponent(state: FacilitatorState): number {
    const samples = state.latencyMs.toArray();
    if (samples.length === 0) return 1.0;
    // p95 inverted, clamped to [500ms, 5000ms].
    const sorted = samples.slice().sort((a, b) => a - b);
    const p95Idx = Math.max(0, Math.floor(sorted.length * 0.95) - 1);
    const p95 = sorted[p95Idx];
    if (p95 <= 500) return 1.0;
    if (p95 >= 5000) return 0.0;
    return 1.0 - (p95 - 500) / 4500;
  }
}
